### DISCUSSION TITLE
Ruby debugging (DAP)

### DISCUSSION THREAD

**Discussion by danillos** [helix-editor/helix#7892] [Q&A]
Someone did https://github.com/ruby/debug works with Helix?

So far I didn't had good results.

Not using Helix, I can start a session doing:

```
rdbg -p 1234 -o -c bundle exec rspec
```

and to connect with
```
rdbg --attach 1234
```

I don't know how to configure it:

```
[language.debugger]
command = "rdbg"
args =["-o", "-c"]
name = "debug"
transport = "tcp"
port-arg = "-p {}"

[[language.debugger.templates]]
name = "launch"
request = "launch"
completion = [ { name = "entrypoint", completion = "filename", default = "bundle exec rspec" } ]
args = { program = "{0}" }

[[language.debugger.templates]]
name = "attach"
request = "attach"
completion = [ "pid" ]
args = { attachCommands = ["rdbg --attach 1234"] }

```

**Comment by ascarter** [comment]
I have not tried Ruby debugger out yet but you likely need to get the Ruby debugger in DAP mode. Take a look at the [VS Code extension](https://github.com/ruby/vscode-rdbg). It mentions setting environment variable `RUBY_DEBUG_DAP_SHOW_PROTOCOL=1` to enable `rdbg` to enable DAP protocol.

**Comment by ascarter** [comment]
I spent some time trying to make this work. I have the following configuration for Ruby LSP, standardrb LSP (for formatting) and rdbg. LSP works pretty good. Debugger - I haven't been successful but I found to at least some information.

`~/.config/helix/languages.toml`:
```
[language-server.ruby-lsp]
command = "ruby-lsp"
config = { provideFormatter = true }

[language-server.standardrb]
command = "standardrb"
args = [ "--lsp" ]
config = { provideFormatter = true }

[[language]]
name = "ruby"
auto-format = true
language-servers = [ { name = "ruby-lsp", except-features = [ "format" ]}, "standardrb" ]

[language.debugger]
name = "rdbg"
transport = "tcp"
command = "rdbg"
args = ["--open"]
port-arg = "--port {}"

[[language.debugger.templates]]
name = "script"
request = "launch"
completion = [ { name = "script", completion = "filename" } ]
args = { program = "{0}" }
```

I had to change `helix-dap/src/client.rs` to not redirect stdin to null for the debugger to start up otherwise it seems to immediately exit after the `spawn` call. However, I'm not convinced there isn't something else going on. I really can only get it to stay up at all if I put Helix in the VS Code debugger. Running the same build from the command line, it seems to immediately exit after spawn. It will now at least stay up long enough for it to try to initiate the session though if in VSCode debugger (using CodeLLDB). But so far I haven't figured out what it is sending to rdbg and what rdbg is responding with. I'll have to try putting rdbg in the debugger to see what it is doing. I'm relatively inexperienced in Rust so debugging from Helix side is slow for me. I know Ruby much better so will try some debugging from the Ruby side next.

```
diff --git a/helix-dap/src/client.rs b/helix-dap/src/client.rs
index ed4515fe..4ee39e6c 100644
--- a/helix-dap/src/client.rs
+++ b/helix-dap/src/client.rs
@@ -169,16 +169,32 @@ pub async fn tcp_process(
     ) -> Result<(Self, UnboundedReceiver<Payload>)> {
         let port = Self::get_port().await.unwrap();
 
-        let process = Command::new(cmd)
-            .args(args)
-            .args(port_format.replace("{}", &port.to_string()).split(' '))
-            // silence messages
-            .stdin(Stdio::null())
+        let command_str = format!(
+            "{} {} {}",
+            cmd,
+            args.join(" "),
+            port_format
+                .replace("{}", &port.to_string())
+                .split(' ')
+                .collect::<Vec<_>>()
+                .join(" ")
+        );
+
+        log::info!("Spawning process: {}", command_str);
+
+        let mut command = Command::new(cmd);
+        command.args(args);
+        command.args(port_format.replace("{}", &port.to_string()).split(' '));
+
+        // silence messages
+        command
+            // .stdin(Stdio::null())
             .stdout(Stdio::null())
-            .stderr(Stdio::null())
-            // Do not kill debug adapter when leaving, it should exit automatically
-            .spawn()?;
-
+            .stderr(Stdio::null());
+            
+        // Do not kill debug adapter when leaving, it should exit automatically
+        let process = command.spawn()?;
+        
         // Wait for adapter to become ready for connection
         time::sleep(time::Duration::from_millis(500)).await;
 
```
